alacritty/src/renderer/shader.rs

```diff
@@ -0,0 +1,161 @@
+        }
+
+        Ok(program)
+    }
+
+    /// Get uniform location by name. Panic if failed.
+    pub fn get_uniform_location(&self, name: &'static CStr) -> Result<GLint, ShaderError> {
+        // This call doesn't require `UseProgram`.
+        let ret = unsafe { gl::GetUniformLocation(self.id(), name.as_ptr()) };
+        if ret == -1 {
+            return Err(ShaderError::Uniform(name));
+        }
+        Ok(ret)
+    }
+
+    /// Get the shader program id.
+    pub fn id(&self) -> GLuint {
+        self.0
+    }
+}
+
+impl Drop for ShaderProgram {
+    fn drop(&mut self) {
+        unsafe { gl::DeleteProgram(self.0) }
+    }
+}
+
+/// A wrapper for a shader id, with automatic lifetime management.
+#[derive(Debug)]
+struct Shader(GLuint);
+
+impl Shader {
+    fn new(kind: GLenum, source: &'static str) -> Result<Self, ShaderError> {
+        let len: [GLint; 1] = [source.len() as GLint];
+
+        let shader = unsafe { Self(gl::CreateShader(kind)) };
+
+        let mut success: GLint = 0;
+        unsafe {
+            gl::ShaderSource(shader.id(), 1, &(source.as_ptr() as *const _), len.as_ptr());
+            gl::CompileShader(shader.id());
+            gl::GetShaderiv(shader.id(), gl::COMPILE_STATUS, &mut success);
+        }
+
+        if success != GLint::from(gl::TRUE) {
+            return Err(ShaderError::Compile(get_shader_info_log(shader.id())));
+        }
+
+        Ok(shader)
+    }
+
+    fn id(&self) -> GLuint {
+        self.0
+    }
+}
+
+impl Drop for Shader {
+    fn drop(&mut self) {
+        unsafe { gl::DeleteShader(self.0) }
+    }
+}
+
+fn get_program_info_log(program: GLuint) -> String {
+    // Get expected log length.
+    let mut max_length: GLint = 0;
+    unsafe {
+        gl::GetProgramiv(program, gl::INFO_LOG_LENGTH, &mut max_length);
+    }
+
+    // Read the info log.
+    let mut actual_length: GLint = 0;
+    let mut buf: Vec<u8> = Vec::with_capacity(max_length as usize);
+    unsafe {
+        gl::GetProgramInfoLog(program, max_length, &mut actual_length, buf.as_mut_ptr() as *mut _);
+    }
+
+    // Build a string.
+    unsafe {
+        buf.set_len(actual_length as usize);
+    }
+
+    // XXX should we expect OpenGL to return garbage?
+    String::from_utf8(buf).unwrap()
+}
+
+fn get_shader_info_log(shader: GLuint) -> String {
+    // Get expected log length.
+    let mut max_length: GLint = 0;
+    unsafe {
+        gl::GetShaderiv(shader, gl::INFO_LOG_LENGTH, &mut max_length);
+    }
+
+    // Read the info log.
+    let mut actual_length: GLint = 0;
+    let mut buf: Vec<u8> = Vec::with_capacity(max_length as usize);
+    unsafe {
+        gl::GetShaderInfoLog(shader, max_length, &mut actual_length, buf.as_mut_ptr() as *mut _);
+    }
+
+    // Build a string.
+    unsafe {
+        buf.set_len(actual_length as usize);
+    }
+
+    // XXX should we expect OpenGL to return garbage?
+    String::from_utf8(buf).unwrap()
```

Comment:
I was looking into that, and can't we use lossy convertion just to remove this unwrap?

Reply:
Pretty sure I didn't mention that because it also exists in the old version, but I don't see why we wouldn't.

Reply:
Done.